Add missing method to list block.

BE-Webdesign opened the pull request:
The list block does not have an on change set for it and is expected to;
causing errors and no updating of the block attribute state.  This PR
copies the current implementation from a text block and enters it into
the list block.

BE-Webdesign committed:
Add missing method to list block.

The list block does not have an on change set for it and is expected to
causing errors and no updating of the block attribute state.  This PR
copies the current implementation from a text block and enters it into
the list block.

nylen commented:
This looks like a good change, but I'm still seeing that this method is not being triggered and any updates are being lost.  I'm also unable to focus the list block if I click within a `li` element - clicking outside one of the item texts works fine.  These two things are probably related.

BE-Webdesign commented:
Yeah good point, it is updating content but the actual items state is not being updated, should the list block keep track of both or should it be unified.  Having two sources of truth kinda makes problems like this easy to miss good catch.

In short:

`attributes.items` is not being kept in sync with `attributes.content`.  Should be easily fixable.

BE-Webdesign commented:
Just realized there should be no content attribute for the list block oops.  Will try and update properly.

BE-Webdesign commented:
Okay rebased and updated, so the redux state for a list block will update properly on change.

youknowriad commented:
I'm still seeing an error on the console (and the state don't update properly) when I have a single `li`  because this doesn't create an LI.

What bothers me a bit is that the block author needs to be aware of the shape of the underlying data. I'm wondering if it simpler if we change the parsing of `items` to `children( 'ol,ul' )`?

BE-Webdesign commented:
>I'm still seeing an error on the console (and the state don't update properly) when I have a single li because this doesn't create an LI.

I didn't test for a single list item, so I will fix that.  Does the data not come as an array?  If not it should be like `[ oneItem ]`, so there is a consistent input output type.  I will check into getting this right, good eyes.

youknowriad commented:
> Does the data not come as an array? 

The data received from `children` call or from the `Editable` is a valid React's `children` prop. It can be an array, an element or a nullable.

BE-Webdesign commented:
Ah then React made a bad choice then lol.

BE-Webdesign commented:
Yeah.... this is super buggy lol.

BE-Webdesign commented:
TinyMCE or something is doing some very bizarre things in the list block.

BE-Webdesign committed:
Fix to account for different react component values.

Adds in conditional logic to account for when there is a list of
elements, a single element, or a null element.

androb commented:
@EphoxJames you might want to look at this

BE-Webdesign commented:
I don't know enough about TinyMCE to know what it is doing.  Basically sometimes the content will be an empty `<br />` with `data-mce-bogus` being passed forward.  From my understanding that is when the data is empty?  It is unclear whether this should be passed forward into the Redux state or what the line of action is exactly.

BE-Webdesign commented:
Just tested and the <br /> is applied to the block state regardless everytime for any block type when it is made empty.  Not sure if that is what we want to have happen.  Going to open a separate issue.

nylen commented:
I just opened #567 which looks potentially related.  However, the changes in this branch **do not** fix the issues I noted there - in fact, the new `onChange` method added in this PR is never called at all for me.

BE-Webdesign commented:
Anything that needs to be changed on this?

nylen commented:
@BE-Webdesign this PR didn't fix the issues noted in the description for me.  They've since been fixed, though I'm not sure exactly where.

Separately, #598 achieved most of what we wanted to do here, so I think we can close this one out.  Please reopen if I missed something.
